### PR TITLE
DM-26973: Updates to WebDAV datastore (token refreshing, performance fixes & more)

### PR DESCRIPTION
**Pull request by bgounon**
This update brings several changes to the WebDAV datastore:
- Implement the possibility to change the bearer token while running, by reading it from a file instead of an environment variable
- Use a shared requests.Session for all instances of ButlerHttpURI to optimize performance (+ add retries on failed requests)
- Better support for parallel ingestion (fix a race condition where different ingestion tasks would try to create the same directory)
- Better support for HTTP redirections (detect if endpoint redirects PUT requests before sending data)
- Make environment variables names more homogeneous
- Implement the possibility to add an "Expect: 100-Continue" header in requests, for servers that support it (implemented mainly for dCache compatibility, may not work as expected with other endpoints).

It also implements new unit tests, docstrings and debug logs.

### CODE REVIEW

File: python/lsst/daf/butler/core/webdavutils.py [hunk 1]
```diff
@@ -50,49 +52,116 @@ def getHttpSession() -> requests.Session:
...
+    The LSST_BUTLER_WEBDAV_CA_BUNDLE must be set to the directory
+    where CA certificates are stored if you intend to use HTTPS
+    to communicate with the endpoint.
+
+    The LSST_BUTLER_WEBDAV_AUTH must be set to obtain a session.
     Depending on the chosen method, additional
     environment variables are required:
 
-    X509: must set WEBDAV_PROXY_CERT
+    X509: must set LSST_BUTLER_WEBDAV_PROXY_CERT
```
Comment (timj) [inline python/lsst/daf/butler/core/webdavutils.py]: Maybe do these as a bulleted list?

File: python/lsst/daf/butler/core/webdavutils.py [hunk 2]
```diff
@@ -50,49 +52,116 @@ def getHttpSession() -> requests.Session:
...
+    (OPTIONAL) LSST_BUTLER_WEBDAV_EXPECT100: if set, we will add an
+    "Expect: 100-Continue" header in all requests. This is required
+    on certain endpoints where requests redirection is made.
 
-    NB: requests will read CA certificates in REQUESTS_CA_BUNDLE
+    NB: requests will read CA certificates in LSST_BUTLER_WEBDAV_CA_BUNDLE
     It must be manually exported according to the system CA directory.
     """
+
+    retries = Retry(total=3, backoff_factor=1, status_forcelist=[429, 500, 502, 503, 504])
```
Comment (timj) [inline python/lsst/daf/butler/core/webdavutils.py:70]: Thanks for adding retry.

File: python/lsst/daf/butler/core/webdavutils.py [hunk 3]
```diff
@@ -50,49 +52,116 @@ def getHttpSession() -> requests.Session:
...
+        refreshToken(session)
     else:
-        raise ValueError("Environment variable WEBDAV_AUTH_METHOD must be set to X509 or TOKEN")
+        raise ValueError("Environment variable LSST_BUTLER_WEBDAV_AUTH must be set to X509 or TOKEN")
+
+    ca_bundle = None
+    try:
+        ca_bundle = os.environ['LSST_BUTLER_WEBDAV_CA_BUNDLE']
+    except KeyError:
+        log.warning("Environment variable LSST_BUTLER_WEBDAV_CA_BUNDLE is not set: \
```
Comment (timj) [inline python/lsst/daf/butler/core/webdavutils.py]: If this is a single sentence the preference at LSST is to write it as:

```python
log.warning("Environment variable LSST_BUTLER_WEBDAV_CA_BUNDLE is not set: "
            "HTTPS requests will fail. If you intend to use HTTPS, please export this variable.")
```
you don't need to use a backslash to continue the strings if you use quotes.

File: python/lsst/daf/butler/core/webdavutils.py [hunk 4]
```diff
@@ -50,49 +52,116 @@ def getHttpSession() -> requests.Session:
...
+
+    Returns
+    -------
+    isTokenAuth : `bool`
+        True if LSST_BUTLER_WEBDAV_AUTH is set to TOKEN, False otherwise.
+    """
+    try:
+        env_auth_method = os.environ['LSST_BUTLER_WEBDAV_AUTH']
+    except KeyError:
+        raise KeyError("Environment variable LSST_BUTLER_WEBDAV_AUTH is not set, \
```
Comment (timj) [inline python/lsst/daf/butler/core/webdavutils.py]: Again, use string continuation as above in log message. Not a backslash within the string.

File: python/lsst/daf/butler/core/webdavutils.py [hunk 5]
```diff
@@ -184,13 +253,43 @@ def isWebdavEndpoint(path: Union[Location, ButlerURI, str]) -> bool:
     isWebdav : `bool`
         True if the endpoint implements Webdav, False if it doesn't.
     """
+    ca_bundle = None
+    try:
+        ca_bundle = os.environ['LSST_BUTLER_WEBDAV_CA_BUNDLE']
+    except KeyError:
+        log.warning("Environment variable LSST_BUTLER_WEBDAV_CA_BUNDLE is not set: \
```
Comment (timj) [inline python/lsst/daf/butler/core/webdavutils.py]: string continuation. I'll stop commenting on these now.

File: python/lsst/daf/butler/core/webdavutils.py [hunk 6]
```diff
@@ -184,13 +253,43 @@ def isWebdavEndpoint(path: Union[Location, ButlerURI, str]) -> bool:
...
+    This is needed when using PUT operations, to avoid starting
+    to send the data to the endpoint, before having to send it again once
+    the 307 redirect response is received, and thus wasting bandwidth.
+
+    Parameters
+    ----------
+    r : `requests.Response`
+        An HTTP response received when requesting the endpoint
+
+    Returns
```
Comment (timj) [inline python/lsst/daf/butler/core/webdavutils.py:274]: Missing `------` under `Returns`.

File: tests/config/testConfigs/webdav/token [hunk 7]
```diff
@@ -0,0 +1 @@
+XXXXXXXX
```
Comment (timj) [inline tests/config/testConfigs/webdav/token]: missing new line? Or is a newline not going to work?

**Review by timj — APPROVED** [review]
Looks good. Some minor comments.

File: tests/config/testConfigs/webdav/token (hunk 7 above)
Comment (bgounon) [reply tests/config/testConfigs/webdav/token]: Will work even with a newline, I trim those when setting the header. Added one here.

**Comment by bgounon** [comment]
@timj your comments are addressed in the new version.
I'm seeing mypy errors in Github Actions which I didn't have a few days earlier though, they are not related to my code and I'm not sure if there is something I should do to get rid of them ?
```
Run mypy python/lsst
python/lsst/daf/butler/core/named.py:424: error: Variable "builtins.NotImplemented" is not valid as a type
python/lsst/daf/butler/core/named.py:424: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
python/lsst/daf/butler/core/named.py:436: error: Variable "builtins.NotImplemented" is not valid as a type
python/lsst/daf/butler/core/named.py:436: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
python/lsst/daf/butler/core/named.py:442: error: Variable "builtins.NotImplemented" is not valid as a type
python/lsst/daf/butler/core/named.py:442: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
python/lsst/daf/butler/core/ddl.py:225: error: Variable "builtins.NotImplemented" is not valid as a type
python/lsst/daf/butler/core/ddl.py:225: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
python/lsst/daf/butler/registry/interfaces/_bridge.py:73: error: Variable "builtins.NotImplemented" is not valid as a type
python/lsst/daf/butler/registry/interfaces/_bridge.py:73: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
Found 5 errors in 3 files (checked 155 source files)
Error: Process completed with exit code 1.
```

**Comment by timj** [comment]
@TallJimbo did mypy get an update again which broke us?

**Comment by TallJimbo** [comment]
Sure seems like it.  First thing I'd try is to use `bool` instead of `Union[bool, NotImplemented]` in case that's now special instead.  If that doesn't work or nobody has time to bother with it I'll look into it after my Ops Bootcamp presentation is done.

**Comment by timj** [comment]
Release notes: https://mypy-lang.blogspot.com/2020/10/mypy-0790-released.html (the new version is already on conda-forge)

**Comment by timj** [comment]
This does fix it:
```diff
$ git diff
diff --git a/python/lsst/daf/butler/core/ddl.py b/python/lsst/daf/butler/core/ddl.py
index 831e3d2c..a7c6497b 100644
--- a/python/lsst/daf/butler/core/ddl.py
+++ b/python/lsst/daf/butler/core/ddl.py
@@ -222,7 +222,7 @@ class FieldSpec:
     doc: Optional[str] = None
     """Documentation for this field."""
 
-    def __eq__(self, other: Any) -> Union[bool, NotImplemented]:
+    def __eq__(self, other: Any) -> bool:
         if isinstance(other, FieldSpec):
             return self.name == other.name
         else:
diff --git a/python/lsst/daf/butler/core/named.py b/python/lsst/daf/butler/core/named.py
index 2b4327f8..955f992b 100644
--- a/python/lsst/daf/butler/core/named.py
+++ b/python/lsst/daf/butler/core/named.py
@@ -421,7 +421,7 @@ class NamedValueSet(NamedValueMutableSet[K]):
     def __repr__(self) -> str:
         return "NamedValueSet({{{}}})".format(", ".join(repr(element) for element in self))
 
-    def __eq__(self, other: Any) -> Union[bool, NotImplemented]:
+    def __eq__(self, other: Any) -> bool:
         if isinstance(other, NamedValueSet):
             return self._dict.keys() == other._dict.keys()
         else:
@@ -433,13 +433,13 @@ class NamedValueSet(NamedValueMutableSet[K]):
     # As per Set's docs, overriding just __le__ and __ge__ for performance will
     # cover the other comparisons, too.
 
-    def __le__(self, other: AbstractSet[K]) -> Union[bool, NotImplemented]:
+    def __le__(self, other: AbstractSet[K]) -> bool:
         if isinstance(other, NamedValueSet):
             return self._dict.keys() <= other._dict.keys()
         else:
             return NotImplemented
 
-    def __ge__(self, other: AbstractSet[K]) -> Union[bool, NotImplemented]:
+    def __ge__(self, other: AbstractSet[K]) -> bool:
         if isinstance(other, NamedValueSet):
             return self._dict.keys() >= other._dict.keys()
         else:
diff --git a/python/lsst/daf/butler/registry/interfaces/_bridge.py b/python/lsst/daf/butler/registry/interfaces/_bridge.py
index 2c4352f0..a37d9d39 100644
--- a/python/lsst/daf/butler/registry/interfaces/_bridge.py
+++ b/python/lsst/daf/butler/registry/interfaces/_bridge.py
@@ -70,7 +70,7 @@ class FakeDatasetRef:
     def __repr__(self) -> str:
         return f"FakeDatasetRef({self.id})"
 
-    def __eq__(self, other: Any) -> Union[bool, NotImplemented]:
+    def __eq__(self, other: Any) -> bool:
         try:
             return self.id == other.id
         except AttributeError:
```

**Comment by timj** [comment]
I've pushed a fix to the branch to see if it fixes it.

**Review by timj — APPROVED** [review]

**Comment by bgounon** [comment]
Merged after confirmation of Jenkins successful execution.

**Pull request merged**